Add filesystem path strategy. Fixes #742

lazka opened the pull request:
A first draft. Feedback welcome.

mulkieran commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,95 @@ def deferred(definition):
+    """A strategy which generates filesystem path values.
+
+    The generated values include everything which the builtin :func:`open`
+    function accepts i.e. which wont lead to :exc:`ValueError` or
```
Just a typo that caught my eye: "wont".

Zac-HD commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,95 @@ def deferred(definition):
+@defines_strategy
+@composite
+def fspaths(draw, allow_pathlike=True):
+    """A strategy which generates filesystem path values.
```
The more I think about the chance of test code operating on paths, the more worried I get - if some downstream user writes a test that opens example paths in write mode, the results will be nasty.

We definitely need a warning that the resulting paths should not be opened (on pain of `atime` disruption in read-only mode, or debris and clobbering in write-mode).  I'd also consider adding an argument `allow_existing=False`, to exclude paths which already exist from being generated.

lazka replied:
The `allow_existing=False` sounds good to me.

Zac-HD commented on tests/cover/test_type_lookup.py:
```diff
@@ -28,7 +28,7 @@
 blacklist = [
     'builds', 'iterables', 'permutations', 'random_module', 'randoms',
-    'runner', 'sampled_from', 'streaming',
+    'runner', 'sampled_from', 'streaming', 'fspaths',
```
This is entirely correct, and makes me want to register a strategy for `os.PathLike`.  In `hypothesis/searchstrategy/types.py`, add the following at line 131 (and `import os` at the top):

    if hasattr(os, 'PathLike'):  # pragma: no branch
        _global_type_lookup[os.PathLike] = st.fspaths(allow_pathlike=False).map(st._PathLike)

lazka replied:
The problem here is that xdist fails due to fspaths returning different types and as a result different tests being created. xdist then complains that one runner has a test which the other one hasn't and stops.

lazka replied:
But I haven't looked to closely. I'll look into what you have proposed.

Zac-HD replied:
To be clear, `fspaths` *really does* belong on the blacklist - it has multiple output types by design.  

The registry addition is unrelated except that this line is what reminded me of it - though the combination disallow-pathlike/map-pathlike I suggest for the registry is specifically to ensure that we always get a pathlike object.

lazka replied:
Ah, got it, thanks.

lazka replied:
I've added the suggested registration now + a test.

I've done the same for pathlib.Path (which only allows str paths) but I haven't used hypothesis with type inference or mypy/annotations so I'm not sure if that is a good idea.

Zac-HD replied:
The `pathlib.Path` support certainly won't *hurt*, but won't allow resolution of the windows or posix path types (because those are subtypes of the generic path type).

The *ideal* would be to extract `_windows_fspaths()` and `_unix_fspaths` in the strategies module (without `@defines_strategy`, but returning `st.one_of(s)`), and then

```
if PY3 and pathlib is not None:  # pragma: no branch
     _global_type_lookup.update({
        pathlib.PurePosixPath: st._unix_fspaths().map(os.fsdecode).map(pathlib.PurePosixPath),
        pathlib.PureWindowsPath: st._windows_fspaths().map(os.fsdecode).map(pathlib.PureWindowsPath),
    })
    if os.name == 'nt':  # pragma: no branch
        _global_type_lookup[pathlib.WindowsPath] = st._windows_fspaths().map(os.fsdecode).map(pathlib.WindowsPath)
    else:
        _global_type_lookup[pathlib.PosixPath] = st._unix_fspaths().map(os.fsdecode).map(pathlib.PosixPath)
```

However I can see this is starting to get out of scope, so feel free to leave it for a later enhancement if you prefer!

lazka replied:
I'm still not convinced that emulating Windows path types on Unix and vice versa is a good idea. I'd like to file that under "leave it for a later enhancement" and would like to see a real use case first.

I've added PurePath support though.

lazka commented on tests/cover/test_fspaths.py:
```diff
@@ -0,0 +1,218 @@
+def test_open(path):
+    # The value range of fspaths() is limited by what open() accepts
+    try:
+        with open(path):
```
open defaults to mode="r", so there shouldn't be any side effects besides things like autofs doing mount on access, which wouldn't change when exists() is used as far as I can see.

I'll try to limit paths to a subdirectory in "/tmp"

lazka replied:
Ah, you mentioned "atime" above, didn't think of that.

mulkieran commented on tests/cover/test_fspaths.py:
```diff
@@ -0,0 +1,217 @@
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
```
Is there any useful test using ```os.path.join()``` or ```pathjoin()```, i.e., can you make some assertion about the properties of the joined path or even just about the operation?

lazka replied:
The only assert I can think of is
```python
type(os.fspath(path)) is type(os.path.join(path, path))
```

Zac-HD replied:
I'd throw it in!

lazka replied:
yeah, added it under test_path_join()

lazka commented:
I think I've addressed all issues raised so far.

Zac-HD commented on tests/cover/test_fspaths.py:
```diff
@@ -0,0 +1,254 @@
+
+
+@pytest.fixture(scope='module')
+def tempdir_path():
```
👍, but I'd delete the `print` and use `try: yield dir_; finally: os.rmdir(dir_)` to ensure the cleanup always happens

lazka replied:
oops, right

Zac-HD commented on tests/cover/test_fspaths.py:
```diff
@@ -0,0 +1,254 @@
+
+@pytest.mark.skipif(not hasattr(os, 'PathLike'), reason='No os.PathLike')
+@given(from_type(getattr(os, 'PathLike', None)))
+def test_registered_strategy_for_type(path):
```
This and the next test are both great, but should be moved to tests/cover/test_type_lookup.py

Zac-HD commented on tests/cover/test_fspaths.py:
```diff
@@ -0,0 +1,254 @@
+
+
+def test_find_basic():
+    find(fspaths(allow_existing=True), lambda p: not fspath(p))
```
find -> example

(though I just discovered that @DRMacIver disagrees with me on this, so it may not be the best advice.  Sorry!)

Zac-HD commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+
+    strategies = []
+
+    if os.name == 'nt':  # pragma: no cover
```
I *think* this should be `pragma: no branch`, because the conditional and other branch will still be executed.

Zac-HD commented on src/hypothesis/searchstrategy/types.py:
```diff
@@ -129,6 +135,19 @@ def from_typing_type(thing):
+# Also check for PY3 because there is a Python 2 pathlib port on pypi
+# which wont work
+if PY3 and pathlib is not None:  # pragma: no branch
+    _global_type_lookup[pathlib.Path] = \
```
> Not convinced that cross-OS support is a good idea, but have added PurePath

Fair enough.  In that case though, it would still be good to resolve to the native concrete path type - if there's a strategy registered for WindowsPath it can also be used for Path, PureWindowsPath, and PurePath as it's a subtype of all those things.  Registering strategies for the parent types is strictly optional (but nice to have).  Eg:

```
path_strings = st.fspaths(allow_pathlike=False).map(os.fsdecode)
if os.name == 'nt':
    os_paths = [pathlib.WindowsPath, pathlib.PureWindowsPath]
else:
    os_paths = [pathlib.PosixPath, pathlib.PurePosixPath]
_global_type_lookup.update({
    t: path_strings.map(t) for t in os_paths + [pathlib.Path, pathlib.PurePath]
})
```

lazka replied:
Can you give an example where that would make a difference compared to the current mapping?

Zac-HD replied:
Sure - try `from_type(pathlib.PosixPath).example()`.  This will only work if a strategy is registered for PosixPath.  Registering something *only* for the child type also works:

```
path_type = pathlib.WindowsPath if os.name == 'nt' else pathlib.PosixPath
_global_type_lookup[path_type] = \
    st.fspaths(allow_pathlike=False).map(os.fsdecode).map(path_type)
```

This will resolve Path and PurePath on all platforms, and the os-specific variants on the relevant operating system.

lazka replied:
Ah, thanks. I got confused by `from_type(pathlib.PosixPath).example()` already returning a PosixPath and assumed that it fell back to the base class strategy which would be the right thing here. But it just falls back to `builds(PosixPath)`..

Zac-HD replied:
Right, the magical `builds()` fallback option kicked in.  Apart from that, all examples in `from_type(T)` should be instances of `T`, which means that we can resolve the parent type to a strategy for child types but not vice-versa.

Zac-HD commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+
+    strategies = []
+
+    if os.name == 'nt':  # pragma: no branch
```
Sorry - looks like this *does* need to be a `pragma: no cover` 😓

lazka replied:
Yeah https://coverage.readthedocs.io/en/coverage-4.3.4/excluding.html#branch-coverage

so "no branch" will exclude just the marked case and imply "no branch" if I understand correctly.

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+
+    :param bool allow_pathlike:
+        If the result can be a pathlike (see :class:`os.PathLike`)
+    :param bool allow_existing:
```
This parameter is a good idea, and I totally agree with its addition, but I wonder if (perhaps in future work? I'm definitely not saying you *need* to do this) it would be worth having `allow_existing` set to True causing it to sometimes explicitly look for existing paths and returning those? Right now the chances of it finding an actually existing file are basically zero unless you have a bunch of single character file names in your working directory.

lazka replied:
Yeah, it's more of a safe guard right now. Sounds like a good future improvement. Random selection of listdir() results and traversing down for example.

lazka commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+    :func:`sys.getfilesystemencoding`.
+
+    :param bool allow_pathlike:
+        If the result can be a pathlike (see :class:`os.PathLike`)
```
Sounds good.

lazka commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+    strategies = []
+
+    if os.name == 'nt':  # pragma: no cover
+        hight_surrogate = characters(
```
yes, typo, thanks.

will add a comment

lazka replied:
(Windows paths can contain surrogates and concatenated paths can contain surrogate pairs, this makes it more likely that that happens)

lazka commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+
+        random = draw(randoms())
+
+        def shuffle_text(t):
```
It allows to generate a sequences of surrogates create by the surrogateescape handler which when encoded and decoded result in a different unicode code point. They can happen when two paths from listdir() get concatenated (one ending with a surrogate, one starting). (This is tested by test_find_text_surrogates_merge_in_bytes_form()).

A pretty special case, but I wanted to cover all possible value ranges which could have an effect.

I'll look into flatmap/permutations and adding a comment.

lazka commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
+
+        strategies.append(unix_path_text.map(shuffle_text))
+
+    main_strategy = one_of(strategies)
```
For my library my motivation was mostly covering all value/type variants and less about the path itself, thus the limited output.

This all sounds good. I'll try to come up with something.

lazka commented:
Added the new changes as an extra commit for now:

* Split the strategy into three: _filename() for just generating
  the name part, _path_root() for generating a path root, such as a
  Windows drive or just /. And finally fspaths() which combines
  everything. This results in generating more real looking paths.
* Make ASCII chars in the name part more likely. On Windows various
  characters are not allowed in paths, and this makes it more likely.
  Also makes the various separators more likely.
* Change allow_pathlike to None/True/False. Where None means generate
  pathlikes when possible, False means no generation and True means
  they will be generated and it will fail if it can't.
* Make the description of allow_existing a bit more neutral to
  allow generation of real paths as a future improvement.
* For allow_pathlike use os.lstat for checking if a file exists instead
  of os.path.exists(). The later returns False for broken symlinks
  while os.lstat detects that case.

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+
+        relative = sep
+        # [drive_letter]:\
+        drive = builds(lambda *x: tp().join(x), char, just(tp(':')), sep)
```
I don't think this is correct. AIUI the valid windows driver letters are the ascii characters A-Z. Other characters are not a valid drive.

lazka replied:
My thinking here was that by making ascii more likely in the _filename() strategy there should a good chance of generating them, but I'll change it to the used range.

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+    :func:`sys.getfilesystemencoding`.
+
+    :param allow_pathlike:
+        By default makes the strategy also include objects implementing the
```
Rather than say "By default" I'd say "If None" as per the other possible argument values.

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+
+    if not allow_existing:
+
+        def check_not_exists(path):
```
Pedant point: I think this is actually checking more than the name implies because of the ValueError case.

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+                os.lstat(path)
+            except OSError:  # pragma: no cover
+                return True
+            except ValueError:  # pragma: no cover
```
Does this mean that we're filtering out too long paths if and only if we have `allow_existing` enabled? That seems counterintuitive and I suspect we always want to filter that.

Would it make more sense to change this method to `check_valid` and have it look something more like:

```python
def check_valid(path):
    try:
         os.lstat(path)
    except OSError:  # pragma: no cover
         return True
    except ValueError:  # pragma: no cover
         return False
    return allow_existing
```

lazka replied:
> Does this mean that we're filtering out too long paths if and only if we have allow_existing enabled?

Yes. Too long paths seemed useful as they can happen when concatenating values in Python. But filtering is fine with me.

I just noticed that not all Python versions throw ValueError, but WindowsError and that the error codes differ between wine and real Windows. That'll require some tests.

lazka replied:
I'm a bit stuck here. os.lstat on Windows takes about 70ms per call here and slows everything down leading to hypothesis complaining. I'll try to find an alternative solution.

Zac-HD replied:
Would using `os.lstat` on Unix and `os.path.exists` on Windows help?

DRMacIver replied:
> Yes. Too long paths seemed useful as they can happen when concatenating values in Python. But filtering is fine with me.

Hmm. I can see that there's a valid use case for technically valid file paths that can never exist. I'm not sure whether always filtering them or never filtering them is best, and am happy to leave that up to you, but I think it needs to be consistent which we do regardless of whether `allow_existing` is set.

> Would using os.lstat on Unix and os.path.exists on Windows help?

I think `os.path.exists` calls `os.stat` under the hood (which is equivalent to `os.lstat`), so it will be just as slow.

> os.lstat on Windows takes about 70ms per call here and slows everything down leading to hypothesis complaining.

Windows file system performance is generally quite bad, but that's *astonishingly* bad. Is that on Wine or on a real Windows?

I did some local testing and the performance I get is much better than that - more on the order of 20 microseconds. I tried the following to see if it would help:

```python
def fast_exists(f):
    try:
        with open(f, 'r'):
             return True
    except OSError:
        # TODO: Need to check if the failure was e.g. a permission error
        return False
```

On my system these were almost identically fast, but I wonder if the one you're having problems on is having to do some weird slow emulation of stat and file opening will be faster?

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+            # os.path.exists returns False for broken symlinks, so use lstat
+            try:
+                os.lstat(path)
+            except OSError:  # pragma: no cover
```
Here's a test that should cover this I think (note I have not run it so it probably contains problems, but the idea should work):

```python

from hypothesis import find
import strategies as st
import os

def test_avoids_existing():
    def fresh_local_file():
        return find(st.fspaths(allow_existing=False), lambda x: os.path.sep not in os.fspath(x))

    f1 = find_fresh_local_file()
    f2 = find_fresh_local_file()
    # checks replay is working correctly: find should find the same
    # thing twice. 
    assert f1 == f2

    # now create the file
    with open(f1, 'w'):
        pass

    # The file now exists, so we are forced to avoid it 
    # and have to generate a new one.
    f3 = find_fresh_local_file()
    assert f3 != f1

```

DRMacIver commented on src/hypothesis/searchstrategy/types.py:
```diff
@@ -129,6 +135,25 @@ def from_typing_type(thing):
+        st.fspaths(allow_pathlike=False).map(st._PathLike)
+
+# Also check for PY3 because there is a Python 2 pathlib port on pypi
+# which wont work
```
Out of curiousity, why won't it work?

lazka replied:
The backported module only works with "str", which works fine on Unix but on Windows you need unicode to support non-ansi paths. But I guess filtering based on the type would work... I'll add that. Only downside is it won't get much testing without the module.

DRMacIver replied:
It would be fine to add an extra build step which installs the module under Python 2 and only runs the file system tests FWIW (it'd probably need adding to both the `appveyor.yml` and as a separate travis job). Your choice whether you care enough to do this though.

Zac-HD commented:
@lazka - are you still interested in working on this, or would you like to hand it over to one of the maintainers for the final tweaks?

lazka commented:
@Zac-HD yeah, sorry, go ahead. Kinda lost my motivation as I only need the initial value/type range part for my projects and things are so slow now that I'd rather use my own strategy.

Zac-HD commented:
No worries - thanks again for your contribution!  Even incomplete, we recognize and appreciate that a lot of work went into this 😄
